ast/parser_test.go

```diff
@@ -1006,71 +1006,102 @@ func TestChainedCall(t *testing.T) {
-	}
-
-	expected1 := MustParseBody(`x = 1; y = 2; z = [i | [x,y] = arr; arr[_] = i]`)
-
-	if !body1.Equal(expected1) {
-		t.Errorf("Expected enclosed body to equal %v but got: %v", expected1, body1)
-	}
-
-	// Check that parser can handle multiple expressions w/o enclosing braces.
-	input2 := `
-		x = 1 ; # comment after semicolon
-		y = 2   # comment without semicolon
-		z = [ i | [x,y] = arr  # comment in comprehension
-				   arr[_] = i]
-	`
-
-	body2, err := ParseBody(input2)
-	if err != nil {
-		t.Fatalf("Unexpected parse error on enclosed body: %v", err)
+	tests := []struct {
+		note  string
+		input string
+		exp   Body
+	}{
+		{
+			note: "three definitions",
+			input: `
+x = 1
+y = 2
+z = [ i | [x,y] = arr
+	arr[_] = i]
+`,
+			exp: MustParseBody(`x = 1; y = 2; z = [i | [x,y] = arr; arr[_] = i]`),
+		},
+		{
+			note: "three definitions, with comments and w/o enclosing braces",
+			input: `
+x = 1 ; # comment after semicolon
+y = 2   # comment without semicolon
+z = [ i | [x,y] = arr  # comment in comprehension
+	arr[_] = i]
+`,
+			exp: MustParseBody(`x = 1; y = 2; z = [i | [x,y] = arr; arr[_] = i]`),
+		},
+		{
+			note:  "array following call w/ whitespace",
+			input: "f(x)\n [1]",
+			exp: NewBody(
+				NewExpr([]*Term{RefTerm(VarTerm("f")), VarTerm("x")}),
+				NewExpr(ArrayTerm(IntNumberTerm(1))),
+			),
+		},
+		{
+			note:  "set following call w/ semicolon",
+			input: "f(x);{1}",
+			exp: NewBody(
+				NewExpr([]*Term{RefTerm(VarTerm("f")), VarTerm("x")}),
+				NewExpr(SetTerm(IntNumberTerm(1))),
+			),
+		},
+		{
+			note:  "array following array w/ whitespace",
+			input: "[1]\n [2]",
+			exp: NewBody(
+				NewExpr(ArrayTerm(IntNumberTerm(1))),
+				NewExpr(ArrayTerm(IntNumberTerm(2))),
+			),
+		},
+		{
+			note:  "array following set w/ whitespace",
+			input: "{1}\n [2]",
+			exp: NewBody(
+				NewExpr(SetTerm(IntNumberTerm(1))),
+				NewExpr(ArrayTerm(IntNumberTerm(2))),
+			),
+		},
+		{
+			note:  "set following call w/ whitespace",
+			input: "f(x)\n {1}",
+			exp: NewBody(
+				NewExpr([]*Term{RefTerm(VarTerm("f")), VarTerm("x")}),
+				NewExpr(SetTerm(IntNumberTerm(1))),
+			),
+		},
+		{
+			note:  "set following ref w/ whitespace",
+			input: "data.p.q\n {1}",
+			exp: NewBody(
+				NewExpr(&Term{Value: MustParseRef("data.p.q")}),
+				NewExpr(SetTerm(IntNumberTerm(1))),
+			),
+		},
+		{
+			note:  "set following variable w/ whitespace",
+			input: "input\n {1}",
+			exp: NewBody(
+				NewExpr(&Term{Value: MustParseRef("input")}),
+				NewExpr(SetTerm(IntNumberTerm(1))),
+			),
+		},
+		{
+			note:  "set following equality w/ whitespace",
+			input: "input = 2 \n {1}",
+			exp: NewBody(
+				Equality.Expr(&Term{Value: MustParseRef("input")}, IntNumberTerm(2)),
+				NewExpr(SetTerm(IntNumberTerm(1))),
+			),
+		},
 	}
 
-	if !body2.Equal(expected1) {
-		t.Errorf("Expected unenclosed body to equal %v but got: %v", expected1, body1)
+	for _, tc := range tests {
```

Comment:
I love that this single test got expanded out to have 10+ testcases. Better coverage is always welcome! :smiley: